### ISSUE TITLE
Wrong dim for torch.cumsum(token_priority) & Identical router_probs / router_logits Output

### ISSUE THREAD

**Issue by wangchaoxixixi** [huggingface/transformers#48293]
https://github.com/huggingface/transformers/blob/550d7b3834670483a4df436541272c055dc364bf/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L102

Is `dim=-3` correct here rather than `-2` for `torch.cumsum`?

`expert_index` has shape `[batch_size, seq_len, 1, num_experts]`; cumulative sum should operate over the `seq_len` dimension.

Also noticed returned `router_probs` and `router_logits` seem to have identical values.

**Comment by poorpaper** [comment]
@wangchaoxixixi 
I can confirm both findings. They are real regressions and are reachable through the normal `SwitchTransformersSparseMLP` execution path.

`SwitchTransformersSparseMLP.forward` flattens the hidden states to `[num_tokens, hidden_dim]` before calling the router:

https://github.com/huggingface/transformers/blob/550d7b3834670483a4df436541272c055dc364bf/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L185-L189

After `torch.max(..., keepdim=True)` and `one_hot`, `expert_index` has shape `[num_tokens, 1, num_experts]`. Consequently, `dim=-2` is the singleton dimension, so every selected entry has priority 1. For any `expert_capacity >= 1`, no token is dropped even when an expert exceeds its capacity.

`dim=-3` addresses the token dimension for both the flattened internal input and the documented `[batch, seq, hidden]` input.

A deterministic reproduction on the current code is:

```python
import torch

from transformers import SwitchTransformersConfig
from transformers.models.switch_transformers.modeling_switch_transformers import (
    SwitchTransformersTop1Router,
)


config = SwitchTransformersConfig(
    d_model=4,
    d_ff=8,
    num_experts=2,
    expert_capacity=2,
    router_jitter_noise=0.0,
)

router = SwitchTransformersTop1Router(config).eval()

with torch.no_grad():
    # Every token deterministically selects expert 0.
    router.classifier.weight.zero_()

# This is the same 2-D shape received by the router from SparseMLP.
hidden_states = torch.ones(4, config.d_model)

returned_probs, expert_mask, returned_logits = router(hidden_states)

print(expert_mask[:, 0, 0].sum().item())
# Current: 4
# Expected: 2

print(torch.equal(returned_probs, returned_logits))
# Current: True

print(returned_logits.shape)
# Current: torch.Size([4, 1])
# Expected raw router-logit shape: torch.Size([4, 2])
```

The identical outputs are also deterministic. The raw classifier output stored in `router_logits` is overwritten by:

```python
router_logits, expert_index = torch.max(
    router_probs, dim=-1, keepdim=True
)
```

The same maximum is then computed again for `router_probs`, so the first and third returned tensors are identical:

https://github.com/huggingface/transformers/blob/550d7b3834670483a4df436541272c055dc364bf/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L96-L107

Both regressions appear to have been introduced by #40132. Before that refactor:

- `argmax` did not retain the extra singleton dimension, so `dim=-2` correctly referred to the token dimension.
- The raw `router_logits` remained separate from the maximum routing probability.

Previous implementation:

https://github.com/huggingface/transformers/blob/e6a8e7debed69d4cf5d493158f3f347931c24c9a/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L203-L215

One caveat for the fix: `SwitchTransformersSparseMLP` currently consumes the third return value as the routing weights, while `OutputRecorder(index=2)` expects that value to contain the raw router logits. Therefore, the router return contract and its caller should be corrected together rather than only renaming the overwritten variable.

I’ve confirmed and reproduced this regression on `transformers==5.15.1`. I’d like to work on a fix covering both the expert-capacity routing and the router-logits return contract. Could you please assign this issue to me?

**Comment by Rocketknight1** [comment]
Yes, I think these are regressions! We generally try to discourage users just running a code agent on random issues though @poorpaper. @wangchaoxixixi you can make a fix if you like, or if not I can open the issue to other contributors

**Comment by wangchaoxixixi** [comment]
@Rocketknight1 Thanks for the feedback. I’m still getting familiar with the Transformers codebase, so I’m not confident enough to provide a proper fix for this issue. It would be better to open this up to other contributors.
